documentation/man/einit-sysvd.8
.\" This manpage has been automatically generated by docbook2man 
.\" from a DocBook document.  This tool can be found at:
.\" <http://shell.ipoline.com/~elmert/comp/docbook2X/> 
.\" Please send any bug reports, improvements, comments, patches, 
.\" etc. to Steve Cheng <steve@ggi-project.org>.
.TH "EINIT-SYSVD" "8" "24 February 2008" "einit-sysvd-0.40.0" ""

.SH NAME
einit-sysvd \- SysV compatibility daemon for eINIT
.SH SYNOPSIS

\fBeinit-sysvd\fR [ \fB-a \fIaddress\fB\fR ]

.SH "DESCRIPTION"
.PP
\fBeinit-sysvd\fR is a small helper daemon that translates certain sysv events into einit events.
It does so by opening /dev/initctl and whenever certain requests are received, it will open a connection to einit
and submit the appropriate commands.
.PP
In particular, you need this daemon if you want to use the 'shutdown', 'reboot', 'halt' or 'poweroff' commands,
or if you want to use the shutdown/reboot options provided in E17, GNOME, KDE, XFCE, MythTV, etc. This daemon will
also translate commands like 'init \fIrunlevel\fR\&' for einit.
.SH "OPTIONS"
.TP
\fB-a \fIaddress\fB\fR
The 9P2000 address to connect to. The default is \fBunix!/dev/einit-9p\fR\&. This daemon will only
connect to this address when an event actually needs to be sent, and it will disconnect again right after that.
.SH "ENVIRONMENT"
.PP
Some command line options may be used to set defaults for this programme.
.TP
\fBEINIT_9P_ADDRESS=\fIaddress\fB\fR
Equivalent to the -a option. The command-line option overrides this variable.
.SH "USAGE"
.PP
Nothing much to see here, run the programme as root and it will detach from the current terminal, write
its pidfile and then just more or less sit there waiting for a connection.
.SH "FILES"
.TP
\fB/dev/einit-9p\fR
The unix socket that is used (by default) for the 9p2000(.u) IPC.
.TP
\fB/var/run/einit-sysvd.pid\fR
This daemon will write its pid to this file.
.SH "BUGS"
.PP
Please send bug reports to the \fBissue tracker on einit.org\fR\&.
.PP
Remember that this is BETA software: avoid using this on mission-critical systems.
.SH "SEE ALSO"
.PP
einit(8), einit-feedback(8), einit-log(8)
.PP
The official project website <URL:http://einit.org/>
.SH "AUTHOR"
.PP
Written by Magnus Deininger <mdeininger@jyujin.de>\&.
.SH "COPYRIGHT"
.PP
(C) 2008 Magnus Deininger, All rights reserved.
.PP
Redistribution and use in source and binary forms, with or without modification,
are permitted provided that the following conditions are met:
.PP
.TP 0.2i
\(bu
Redistributions of source code must retain the above copyright notice,
this list of conditions and the following disclaimer.
.TP 0.2i
\(bu
Redistributions in binary form must reproduce the above copyright notice,
this list of conditions and the following disclaimer in the documentation
and/or other materials provided with the distribution.
.TP 0.2i
\(bu
Neither the name of the project nor the names of its contributors may be
used to endorse or promote products derived from this software without
specific prior written permission.
.PP
THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


ChangeLog
2008/2/18: Ryan Hope
 ? initramfs support
 * timezone support
 ? urandom
 * bootchart support
 * curses feedback daemon
 * better guile support
 * more scheme modules
 * network-v2
 * module-logic-v4
 * logd
 * ipc-9p

2007/11/26: Ryan Hope
 * scheme modules
 * edev
 * switched to git
 * hotplug handler

2007/04/30: mdeininger
 * enhanced feedback
 * fixed some deadlocks

2007/04/24: mdeininger
 i'm getting quite lazy about updating this file... have a look at the svn
 revision history for closer details ;)
 * implemented shutdown-(before|after)
 * fixed problems with before/after (should work again)

2007/04/15: mdeininger
 * more API changes...

2007/04/14: mdeininger
 * API changes... lots of 'em...

2007/04/12: mdeininger
 * changes to the feedback and feedback auto-add code
 * new feedback module for festival (gonna use a thread like with fbsplash)

2007/04/11: mdeininger
 * a number of minor fixes
 * the new core seems to be doing quite well

2007/04/10: mdeininger
 * new core mostly operational.
 * changed configure script to support modules in C++
 * added template for C++ module (ipc-dbus)

2007/04/09: mdeininger
 * started on a new more-or-less event-based module logics core

2007/04/08: mdeininger
 * made suff cleaner by removing the 'feedback' mode, but also auto-adding
   a feedback module if none is specified

2007/04/05: mdeininger
 * added options="feedback" to daemon/shell modules to tag them as feedback-
   modules (gonna be useful for the fbsplash module)
 * moved re-mounting the root-filesystem to mount/critical
 * the fbsplash module seems to work just fine =)
 * moved the control socket to /dev/einit-control
 * removed module reset/reload/zap hooks in favour of "custom" hooks
 * changed mount/* to mount-

2007/04/03: mdeininger
 * parsed some of the configuration in the wrong function/forgot updates
 * "list services" has been moved to the module-logics core, + it will now
   display candidates in the order that einit will try to deal with them.
 * added envfile="" support
 * added is-up="" and is-down="" attrs to daemons -- they run after the daemon
   is forked -- if the return value is shell-ish true, then the appropriate
   change to the daemon is considered successful, else it is seen as failed
 * added need-files="" to daemons -- just add a list of files that must exist
   or else einit won't even try to bring it up (if the list is available).
 * new module: configuration-network

2007/04/02: mdeininger
 * need to split that configuration up :D
 * use /etc/fstab by default
 * introduced a new readdirfilter() function, we did do that a lot...

2007/04/01: mdeininger
 * started on a custom set of (m|re|c)alloc() functions
 * modified configure script to add files to link into the core

[2006/04/01: mdeininger: release, Version 0.21.0]

[2006/03/27: mdeininger: release, Version 0.20.0]

2007/03/22: mdeininger
 * logging works
 * output is a little clearer

2007/03/21: mdeininger
 * logging kinda works...

2007/03/20: mdeininger
 * stripped einit-control even further

2007/03/19: mdeininger
 * fixed/enhanced the build system
 * added some very simple hashing to the tree-linear code...
   seems to decrease execution time considerably

2007/03/18: mdeininger
 * some good progress with compiling modules into the core

2007/03/14: mdeininger
 * need to be able to compile modules in.

2007/03/13: mdeininger
 * this should get us more "const-correct" :)
 * fixed some annoying compiler warnings
 * moved /etc/einit-control to /etc/einit/control

2007/03/11: mdeininger
 * introduced a new function to check for string matches (strmatch())

2007/03/09: mdeininger
 * added support for ryan's shadow modules

2007/03/07: mdeininger
 * fixes, cleanups, new apach2 module

2007/03/05: mdeininger
 * some new modules, random fixes

2007/03/01: mdeininger
 * the v3 module logics core is now by default external
 * the default expat configuration parser is now external as well.

2007/02/28: mdeininger
 * more preparations to get more stuff out of the core.

2007/02/27: mdeininger
 * shell-gdm should also provide "xdm"
 * the v3 core is now the default
 * fixed some deadlocks in the v3 module-logics core
 * pexec() seems to be less flakey now

2007/02/25: mdeininger
 * some new macros
 * new module logics core is coming along nicely

2007/02/22: mdeininger
 * fixed readfile() problem with .keep files

2007/02/21: mdeininger
 * new new module logics core, again

2007/02/20: mdeininger
 * added service transformations
 * build system changes
 * moved get_plan_progress from the visualiser to the module-logic.c file
 * new sobexsrv module by RedIpS -- nice man ;)

2007/02/19: mdeininger
 * etc/einit/einit.xml has been moved to lib/einit/einit.xml
   (einit.xml should not be modified by users)
 * lib/einit/*.so have been moved to lib/einit/modules
 * added code to read and use the sysctl.conf file
 * ha, found the bloody bugger
 * the xml configuration parser may now create missing include directories
   itself
 * new ipc command "set variable"
 * new ipc command "save configuration"

2007/02/18: mdeininger
 * new function for error reporting (and optional hara-kiri)
 * workover using -Wall
 * added some code to check for errors (that normally need not be checked)

2007/02/16: mdeininger
 * fixed some minor things

2007/02/15: mdeininger
 * extra module for entranceD (additional to the entrance module from earlier)
 * "list services" will now list deprecated modules with a red "-" instead of
   the usual *
 * fixed a stray malloc()

2007/02/08: mdeininger
 * new module for ipc commands that modify the configuration

2007/02/07: mdeininger
 * output for aforementioned ipc call

2007/02/06: mdeininger
 * ipc call to get configuration nodes

[2006/02/05: mdeininger: release, Version 0.16.0]

2007/02/04: mdeininger
 * added a cfg_filter() function

2007/02/02: mdeininger
 * added "rc zap" command
 * einit-control should now auto-add --ansi if appropriate

2007/02/01: mdeininger

src/tools/einit-sysvd.c
/*
 *  einit-sysvd.c
 *  einit
 *
 *  Created by Magnus Deininger on 23/02/2008.
 *  Copyright 2008 Magnus Deininger. All rights reserved.
 *
 */

/*
Copyright (c) 2008, Magnus Deininger
All rights reserved.

Redistribution and use in source and binary forms, with or without modification,
are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright notice,
      this list of conditions and the following disclaimer.
    * Redistributions in binary form must reproduce the above copyright notice,
      this list of conditions and the following disclaimer in the documentation
      and/or other materials provided with the distribution.
    * Neither the name of the project nor the names of its contributors may be
      used to endorse or promote products derived from this software without
      specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

#define _BSD_SOURCE

#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>
#include <sys/param.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <string.h>
#include <fcntl.h>
#include <errno.h>

#include <einit/einit.h>
#include <syslog.h>

#define INITCTL_MAGIC 0x03091969
#define INITCTL_CMD_START        0x00000000
#define INITCTL_CMD_RUNLVL       0x00000001
#define INITCTL_CMD_POWERFAIL    0x00000002
#define INITCTL_CMD_POWERFAILNOW 0x00000003
#define INITCTL_CMD_POWEROK      0x00000004

#define INITCTL_CMD_SETENV       0x00000006
#define INITCTL_CMD_UNSETENV     0x00000007

struct init_command {
 uint32_t signature;    // signature, must be INITCTL_MAGIC
 uint32_t command;      // the request ID
 uint32_t runlevel;     // the runlevel argument
 uint32_t timeout;      // time between TERM and KILL
 char     padding[368]; // padding, legacy applications expect the struct to be 384 bytes long
};

#define PIDFILE "/var/run/einit-sysvd.pid"

#define INITCTL_FIFO "/dev/initctl"
#define INITCTL_FIFO_MODE 0600

#define RUNLEVEL_0 "power-down"
#define RUNLEVEL_1 "single"
#define RUNLEVEL_2 "boot"
#define RUNLEVEL_3 "default"
#define RUNLEVEL_4 "default"
#define RUNLEVEL_5 "default"
#define RUNLEVEL_6 "power-reset"
#define RUNLEVEL_S "single"

char **sysvd_argv = NULL;
int sysvd_argc = 0;

void connect_or_terminate () {
 if (!einit_connect (&sysvd_argc, sysvd_argv)) {
  syslog (LOG_CRIT, "could not connect to einit: %m");
 }
}

int initctl_wait (char *fifo) {
 int nfd;

 while ((nfd = open (fifo, O_RDONLY))) {
  struct init_command ic;

  if (nfd == -1) { /* open returning -1 is very bad, terminate */
   return EXIT_FAILURE;
  }

  memset (&ic, 0, sizeof (struct init_command)); // clear this struct, just in case

  if (read (nfd, &ic, sizeof(struct init_command)) > 12) { // enough bytes to process were read, we dont care about the rest anyway
   if (ic.signature == INITCTL_MAGIC) {
//  INITCTL_CMD_START: what's that do?
//  INITCTL_CMD_UNSETENV is deliberately ignored
    if (ic.command == INITCTL_CMD_RUNLVL) { // switch runlevels (modes...)
     char tmp[BUFFERSIZE], *nmode = NULL;

     switch (ic.runlevel) {
      case '0': nmode = RUNLEVEL_0; break;
      case '1': nmode = RUNLEVEL_1; break;
      case '2': nmode = RUNLEVEL_2; break;
      case '3': nmode = RUNLEVEL_3; break;
      case '4': nmode = RUNLEVEL_4; break;
      case '5': nmode = RUNLEVEL_5; break;
      case '6': nmode = RUNLEVEL_6; break;
      case 's':
      case 'S': nmode = RUNLEVEL_S; break;
      default:
       syslog (LOG_WARNING, "Invalid Runlevel: '%c'", ic.runlevel);
       break;
     }

     if (nmode) {
      syslog (LOG_NOTICE, "initctl: switching to mode %s (runlevel %c)", nmode, ic.runlevel);

      connect_or_terminate();
      einit_switch_mode(nmode);
      einit_disconnect();
     }
    } else if (ic.command == INITCTL_CMD_SETENV) { // padding contains the new environment string
     char **cx = str2set (':', ic.padding);
     if (cx) {
      if (cx[0] && cx[1]) {
       if (strmatch (cx[0], "INIT_HALT")) {
        if (strmatch (cx[1], "HALT") || strmatch (cx[1], "POWERDOWN")) {
         connect_or_terminate();
         einit_power_down();
         einit_disconnect();
        }
       }
      }

      efree (cx);
     }
    } else
     syslog (LOG_WARNING, "invalid initctl received: unknown command");
   } else {
    syslog (LOG_WARNING, "invalid initctl received: invalid signature");
   }
  }

  close (nfd);
 }

 return EXIT_SUCCESS; /* not reached... usually */
}

int main (int argc, char **argv) {
 FILE *pidfile;

 sysvd_argv = argv;
 sysvd_argc = argc;

 unlink (INITCTL_FIFO); /* unlink fifo, ignore if it didn't work */
 if (mkfifo (INITCTL_FIFO, INITCTL_FIFO_MODE) && (errno != EEXIST)) {
  perror ("could not create initctl fifo");
  return EXIT_FAILURE;
 }

 if (daemon(0, 0)) {
  perror ("could not daemonise");
  return EXIT_FAILURE;
 }

 pidfile = fopen (PIDFILE, "w");
 if (pidfile) {
  fprintf (pidfile, "%d\n", getpid());
  fclose (pidfile);
 }

 return initctl_wait(INITCTL_FIFO);
}
